README.md
[DISE Dashboard](http://dise.klp.org.in)
===============

Dashboard for Dise data to allow interactive dise data gathering

New OLAP API Alpha Endpoint
---
Base URL: `http://host:port/api/v1/olap/`

Obtain and restore the database and point to it from local settings file to test.

Accepted parameters:

| name     | required |  description                    |
| -------- | ---------| ------------                    |
| method   |    yes   | Name of one of the available mathods  |
| session  |    yes   | Educational year (Available: `10-11` & `11-12`, Default `10-11`)  |

And you may also pass the parameters of respective method in the url.

Methods
---
 - School
    - [School.getInfo](#schoolgetinfo)
    - [School.search](#schoolsearch)
 - Cluster
    - [Cluster.getSchools](#clustergetschools)
    - [Cluster.search](#clustersearch)

Methods: `School`
---
### `School.getInfo`

Accepted parameters:

| name   | required |  description                    |
| ------ | ---------| ------------                    |
| code   |    yes   | DISE Code                       |

E.g. -

    http://local.dise.klp.org.in:8000/api/v1/olap/?method=School.getInfo&code=29010200101&session=10-11

### `School.search`

Accepted parameters:

| name    | required |  description                    |
| ------  | ---------| ------------                    |
| name    | no | full or part of the name |
| limit   | no | maximum number of results to return |
| geo     | no | if the school has geo location, `true` or `false` |
| cluster   | no | name of cluster to search within |

E.g. To find maximum 50 schools in Mahagaon cluster that has no geolocation and have `govt` as part of their name -

    http://local.dise.klp.org.in:8000/api/v1/olap/?method=School.search&session=10-11&limit=50&cluster=MAHAGAON&geo=false&name=govt


Methods: `Cluster`
---
### `Cluster.getSchools`

Accepted parameters:

| name   | required |  description                    |
| ------ | ---------| ------------                    |
| name   | yes | cluster name |

E.g. -

    http://local.dise.klp.org.in:8000/api/v1/olap/?method=Cluster.getSchools&session=10-11&name=BARAGAON


### `Cluster.search`

Accepted parameters:

| name   | required |  description                    |
| ------ | ---------| ------------                    |
| name   | no | name of cluster to search for |
| block   | no | name of block to search for the cluster |

E.g. -

    http://local.dise.klp.org.in:8000/api/v1/olap/?method=Cluster.search&session=10-11&name=DONGAON


Backing up and restoring database
---
This is what I do as of now. For backing up -

    pg_dump -Ox -h localhost -U klp klpdise_olap > klp/dise/backups/klpdise_olap-2013-11-08-0241-full.pgsql

For restoring the same -

    sudo psql -U klp -h localhost klpdise_olap < klpdise_olap-2013-11-08-0241-full.pgsql

Indicators
---

1. General
    - Rural
    - Urban
    - Government
    - Private

2. Facility (negatives only)
    - Without Pucca Building
    - Classrooms needing repair
    - Without Toilets
    - Without Girls Toilets
    - Without Electricity
    - Without Secure Wall
    - Without Mid-Day-Meal
    - Without Drinking Water
    - Without Medical Checkup
    - Without Library
    - Without Blackboard
    - Without Playground
    - Without Ramps

3. Teachers
    - Without separate room for HM
    - Without HM

4. RTE
    - Without SDMC Constituted
    - Without SDMC Meetings
    - Without Text Books
    - With Children from Weaker Section Enrolled (PVT only)

5. Compound Indicators
    - With enrolment < 25
    - With PTR > 35
    - With Girls:Boys Gender ratio < 1

6. Range Indicators
    - With the number of class rooms between [___] and [___]
    - With the number of Teacher between [___] and [___]
    - With number of girls in class [____] between [____] and [____]
    - With number of boys in class [____] between [____] and [____]


CHANGELOG.md

#### [Current]
 * [371b29f](../../commit/371b29f) - __(Bibhas)__ test commit
 * [600eb26](../../commit/600eb26) - __(Bibhas)__ added changelog

#### v2.1
 * [cdc9812](../../commit/cdc9812) - __(Bibhas)__ Merge pull request [#125](../../issues/125) from klpdotorg/dev

Fixed MOE and School counts on aggregated views, added new columns
 * [9dd48df](../../commit/9dd48df) - __(Bibhas)__ sorting moe and categories
 * [72dce3f](../../commit/72dce3f) - __(Bibhas)__ added title to PTR header
 * [52c84b7](../../commit/52c84b7) - __(Bibhas)__ fixed moe display on front end
 * [74f5cbe](../../commit/74f5cbe) - __(Bibhas)__ reading default session from settings.py
 * [2f50057](../../commit/2f50057) - __(Bibhas)__ returning moe and categories from serializer
 * [bd34b7c](../../commit/bd34b7c) - __(Bibhas)__ added new columns and returning all child schools from aggregation models
 * [9487d72](../../commit/9487d72) - __(Bibhas)__ added 14-15 as default session in settings.py
 * [38d2e6a](../../commit/38d2e6a) - __(Bibhas)__ updated doc for search_choices_by_key
 * [3dd765e](../../commit/3dd765e) - __(Bibhas)__ removed moe from aggregation sqls

#### v2.0
 * [d1a34b6](../../commit/d1a34b6) - __(Bibhas)__ updated tile url to use https
 * [c19f1d2](../../commit/c19f1d2) - __(Bibhas)__ loading fonts via https
 * [b6631e8](../../commit/b6631e8) - __(Bibhas)__ Merge pull request [#124](../../issues/124) from klpdotorg/dev

Updated import scripts, imported 14-15 data, using DRF and removed old codes
 * [fa6391a](../../commit/fa6391a) - __(Bibhas)__ added some tests
 * [68d6a90](../../commit/68d6a90) - __(Bibhas)__ added django-rest-swagger and removed some more old code
 * [0f6901e](../../commit/0f6901e) - __(Bibhas)__ fixed getting info for a single school
 * [965117f](../../commit/965117f) - __(Bibhas)__ Merge pull request [#123](../../issues/123) from klpdotorg/cleanup_old_code

Removed all old code and made DRF code work. 
 * [b02d5d5](../../commit/b02d5d5) - __(Bibhas)__ fixed issues with /schools endpoints
 * [4ffe4e8](../../commit/4ffe4e8) - __(Bibhas)__ Merge pull request [#122](../../issues/122) from klpdotorg/bibhas

Updated DISE data import script and made 14-15 data default
 * [ea120a7](../../commit/ea120a7) - __(Bibhas)__ made 14-15 data the default on UI
 * [f6b091d](../../commit/f6b091d) - __(Bibhas)__ updated aggregation sqls
 * [e93e8f0](../../commit/e93e8f0) - __(Bibhas)__ deleted old import codes
 * [93c3a60](../../commit/93c3a60) - __(Bibhas)__ removed all old code and remaned all files
 * [a744f23](../../commit/a744f23) - __(Bibhas)__ added rte, enrolment and teacher data import script
 * [78d6a9e](../../commit/78d6a9e) - __(Bibhas)__ added facility data import script
 * [ebcb6d4](../../commit/ebcb6d4) - __(Bibhas)__ updated general data import script
 * [fb998bc](../../commit/fb998bc) - __(Bibhas)__ merged
 * [45be6b5](../../commit/45be6b5) - __(Bibhas)__ added 1415 cleanup sql
 * [2b26787](../../commit/2b26787) - __(Bibhas)__ preparing to import 14-15 data
 * [99fd442](../../commit/99fd442) - __(Bibhas)__ Merge pull request [#121](../../issues/121) from klpdotorg/dev

Changes to the search endpoint
 * [0d325d9](../../commit/0d325d9) - __(Bibhas)__ Merge pull request [#120](../../issues/120) from klpdotorg/bibhas

Handling WSGI Error
 * [dd5050d](../../commit/dd5050d) - __(Bibhas)__ handling wsgi changes again
 * [6d844ea](../../commit/6d844ea) - __(Bibhas)__ reverting changes
 * [64ec679](../../commit/64ec679) - __(Bibhas)__ handling wsgi error
 * [1e88101](../../commit/1e88101) - __(Bibhas)__ Merge pull request [#119](../../issues/119) from klpdotorg/bibhas

Pending changes for dev
 * [6dde607](../../commit/6dde607) - __(Bibhas)__ added support for selective search for entities
 * [0f6fb8d](../../commit/0f6fb8d) - __(Bibhas)__ added django sql explorer
 * [84065ba](../../commit/84065ba) - __(Bibhas)__ Merge pull request [#118](../../issues/118) from klpdotorg/bibhas

updated leaflet css location
 * [5fc35f4](../../commit/5fc35f4) - __(Bibhas)__ updated leaflet css location
 * [c54a9eb](../../commit/c54a9eb) - __(Bibhas)__ updated leaflet js url
 * [5ce541f](../../commit/5ce541f) - __(Bibhas)__ Merge pull request [#117](../../issues/117) from klpdotorg/bibhas

Integrating DRF 3.1 and Django 1.7.8
 * [bf69a27](../../commit/bf69a27) - __(Bibhas)__ commented some console logs

apps/schools/models.py
# This is an auto-generated Django model module.
# You'll have to do the following manually to clean this up:
#     * Rearrange models' order
#     * Make sure each model has one field with primary_key=True
# Feel free to rename the models, but don't rename db_table values or field names.
#
# Also note: You'll have to insert the output of 'django-admin.py sqlcustom [appname]'
# into your database.
from __future__ import unicode_literals

from common.models import (
    search_choices, search_choices_by_key, YESNO, AREA, SCHOOL_CATEGORY,
    SCHOOL_MANAGEMENT, SCHOOL_TYPES, MEDIUM, MDM_STATUS, KITCHENSHED_STATUS,
    BOUNDARY_WALL, BUILDING_STATUS, DRINKING_WATER
)

from django.contrib.gis.db import models
from django.db import cached_property
from jsonfield import JSONField
from django.conf import settings
import collections


class BasicData(models.Model):
    school_code = models.BigIntegerField(primary_key=True)
    centroid = models.GeometryField(blank=True, null=True)
    district = models.CharField(max_length=50, blank=True)
    school_name = models.CharField(max_length=200, blank=True)
    block_name = models.CharField(max_length=50, blank=True)
    cluster_name = models.CharField(max_length=50, blank=True)
    village_name = models.CharField(max_length=50, blank=True)
    assembly_name = models.CharField(max_length=35, blank=True)
    parliament_name = models.CharField(max_length=35, blank=True)
    pincode = models.IntegerField(null=True, blank=True)
    new_pincode = models.IntegerField(null=True, blank=True)
    rural_urban = models.IntegerField(choices=AREA, null=True, blank=True)
    medium_of_instruction = models.IntegerField(choices=MEDIUM, null=True, blank=True)
    distance_brc = models.FloatField(null=True, blank=True)
    distance_crc = models.FloatField(null=True, blank=True)
    yeur_estd = models.IntegerField(null=True, blank=True)
    pre_pry_yn = models.IntegerField(null=True, blank=True)
    residential_sch_yn = models.IntegerField(null=True, blank=True)
    sch_management = models.IntegerField(choices=SCHOOL_MANAGEMENT, null=True, blank=True)
    lowest_class = models.IntegerField(null=True, blank=True)
    highest_class = models.IntegerField(null=True, blank=True)
    sch_category = models.IntegerField(choices=SCHOOL_CATEGORY, null=True, blank=True)
    pre_pry_students = models.IntegerField(null=True, blank=True)
    school_type = models.IntegerField(choices=SCHOOL_TYPES, null=True, blank=True)
    shift_school_yn = models.IntegerField(null=True, blank=True)
    no_of_working_days = models.IntegerField(null=True, blank=True)
    no_of_acad_inspection = models.IntegerField(null=True, blank=True)
    residential_sch_type = models.IntegerField(null=True, blank=True)
    pre_pry_teachers = models.IntegerField(null=True, blank=True)
    visits_by_brc = models.IntegerField(null=True, blank=True)
    visits_by_crc = models.IntegerField(null=True, blank=True)
    school_dev_grant_recd = models.FloatField(null=True, blank=True)
    school_dev_grant_expnd = models.FloatField(null=True, blank=True)
    tlm_grant_recd = models.FloatField(null=True, blank=True)
    tlm_grant_expnd = models.FloatField(null=True, blank=True)
    funds_from_students_recd = models.FloatField(null=True, blank=True)
    funds_from_students_expnd = models.FloatField(null=True, blank=True)
    building_status = models.IntegerField(choices=BUILDING_STATUS, null=True, blank=True)
    tot_clrooms = models.IntegerField(null=True, blank=True)
    classrooms_in_good_condition = models.IntegerField(null=True, blank=True)
    classrooms_require_major_repair = models.IntegerField(choices=YESNO, null=True, blank=True)
    classrooms_require_minor_repair = models.IntegerField(choices=YESNO, null=True, blank=True)
    other_rooms_in_good_cond = models.IntegerField(null=True, blank=True)
    other_rooms_need_major_rep = models.IntegerField(null=True, blank=True)
    other_rooms_need_minor_rep = models.IntegerField(null=True, blank=True)
    toilet_common = models.IntegerField(null=True, blank=True)
    toilet_boys = models.IntegerField(null=True, blank=True)
    toilet_girls = models.IntegerField(null=True, blank=True)
    kitchen_devices_grant = models.IntegerField(null=True, blank=True)
    status_of_mdm = models.IntegerField(choices=MDM_STATUS, null=True, blank=True)
    computer_aided_learnin_lab = models.IntegerField(choices=YESNO, null=True, blank=True)
    separate_room_for_headmaster = models.IntegerField(choices=YESNO, null=True, blank=True)
    electricity = models.IntegerField(choices=YESNO, null=True, blank=True)
    boundary_wall = models.IntegerField(choices=BOUNDARY_WALL, null=True, blank=True)
    library_yn = models.IntegerField(choices=YESNO, null=True, blank=True)
    playground = models.IntegerField(choices=YESNO, null=True, blank=True)
    blackboard = models.IntegerField(null=True, blank=True)
    books_in_library = models.IntegerField(null=True, blank=True)
    drinking_water = models.IntegerField(choices=DRINKING_WATER, null=True, blank=True)
    medical_checkup = models.IntegerField(choices=YESNO, null=True, blank=True)
    ramps = models.IntegerField(choices=YESNO, null=True, blank=True)
    no_of_computers = models.IntegerField(null=True, blank=True)
    male_tch = models.IntegerField(null=True, blank=True)
    female_tch = models.IntegerField(null=True, blank=True)
    noresp_tch = models.IntegerField(null=True, blank=True)
    head_teacher = models.IntegerField(null=True, blank=True)
    graduate_teachers = models.IntegerField(null=True, blank=True)
    tch_with_professional_qualification = models.IntegerField(null=True, blank=True)
    days_involved_in_non_tch_assgn = models.IntegerField(null=True, blank=True)
    teachers_involved_in_non_tch_assgn = models.IntegerField(null=True, blank=True)
    class1_total_enr_boys = models.IntegerField(blank=True, null=True)
    class2_total_enr_boys = models.IntegerField(blank=True, null=True)
    class3_total_enr_boys = models.IntegerField(blank=True, null=True)
    class4_total_enr_boys = models.IntegerField(blank=True, null=True)
    class5_total_enr_boys = models.IntegerField(blank=True, null=True)
    class6_total_enr_boys = models.IntegerField(blank=True, null=True)
    class7_total_enr_boys = models.IntegerField(blank=True, null=True)
    class8_total_enr_boys = models.IntegerField(blank=True, null=True)
    class1_total_enr_girls = models.IntegerField(blank=True, null=True)
    class2_total_enr_girls = models.IntegerField(blank=True, null=True)
    class3_total_enr_girls = models.IntegerField(blank=True, null=True)
    class4_total_enr_girls = models.IntegerField(blank=True, null=True)
    class5_total_enr_girls = models.IntegerField(blank=True, null=True)
    class6_total_enr_girls = models.IntegerField(blank=True, null=True)
    class7_total_enr_girls = models.IntegerField(blank=True, null=True)
    class8_total_enr_girls = models.IntegerField(blank=True, null=True)
    total_boys = models.IntegerField(blank=True, null=True)
    total_girls = models.IntegerField(blank=True, null=True)

    objects = models.GeoManager()

    class Meta:
        abstract = True

    def __unicode__(self):
        return self.school_name

    @property
    def popup_content(self):
        return self.school_name


class AggregationBase(models.Model):
    slug = models.CharField(max_length=50, primary_key=True)
    centroid = models.GeometryField(blank=True, null=True)

    sum_schools = models.BigIntegerField(null=True, blank=True)
    sum_govt_schools = models.BigIntegerField(null=True, blank=True)
    sum_rural_schools = models.BigIntegerField(null=True, blank=True)
    avg_distance_brc = models.FloatField(null=True, blank=True)
    avg_distance_crc = models.FloatField(null=True, blank=True)
    sum_pre_primary_schools = models.BigIntegerField(null=True, blank=True)
    sum_residential_schools = models.BigIntegerField(null=True, blank=True)
    sum_pre_primary_students = models.BigIntegerField(null=True, blank=True)
    avg_pre_primary_students = models.FloatField(null=True, blank=True)
    sum_shift_schools = models.BigIntegerField(null=True, blank=True)
    sum_no_of_working_days = models.BigIntegerField(null=True, blank=True)
    avg_no_of_working_days = models.FloatField(null=True, blank=True)
    sum_no_of_acad_inspection = models.BigIntegerField(null=True, blank=True)
    avg_no_of_acad_inspection = models.FloatField(null=True, blank=True)
    sum_visits_by_brc = models.BigIntegerField(null=True, blank=True)
    avg_visits_by_brc = models.FloatField(null=True, blank=True)
    sum_visits_by_crc = models.BigIntegerField(null=True, blank=True)
    avg_visits_by_crc = models.FloatField(null=True, blank=True)
    sum_school_dev_grant_recd = models.FloatField(null=True, blank=True)
    avg_school_dev_grant_recd = models.FloatField(null=True, blank=True)
    sum_school_dev_grant_expnd = models.FloatField(null=True, blank=True)
    avg_school_dev_grant_expnd = models.FloatField(null=True, blank=True)
    sum_tlm_grant_recd = models.FloatField(null=True, blank=True)
    avg_tlm_grant_recd = models.FloatField(null=True, blank=True)
    sum_tlm_grant_expnd = models.FloatField(null=True, blank=True)
    avg_tlm_grant_expnd = models.FloatField(null=True, blank=True)
    sum_funds_from_students_recd = models.FloatField(null=True, blank=True)
    avg_funds_from_students_recd = models.FloatField(null=True, blank=True)
    sum_funds_from_students_expnd = models.FloatField(null=True, blank=True)
    avg_funds_from_students_expnd = models.FloatField(null=True, blank=True)
    sum_tot_clrooms = models.BigIntegerField(null=True, blank=True)
    avg_tot_clrooms = models.FloatField(null=True, blank=True)
    sum_classrooms_in_good_condition = models.BigIntegerField(null=True, blank=True)
    avg_classrooms_in_good_condition = models.FloatField(null=True, blank=True)
    sum_classrooms_require_major_repair = models.BigIntegerField(null=True, blank=True)
    avg_classrooms_require_major_repair = models.FloatField(null=True, blank=True)
    sum_classrooms_require_minor_repair = models.BigIntegerField(null=True, blank=True)
    avg_classrooms_require_minor_repair = models.FloatField(null=True, blank=True)
    sum_other_rooms_in_good_cond = models.BigIntegerField(null=True, blank=True)
    avg_other_rooms_in_good_cond = models.FloatField(null=True, blank=True)
    sum_other_rooms_need_major_rep = models.BigIntegerField(null=True, blank=True)
    avg_other_rooms_need_major_rep = models.FloatField(null=True, blank=True)
    sum_other_rooms_need_minor_rep = models.BigIntegerField(null=True, blank=True)
    avg_other_rooms_need_minor_rep = models.FloatField(null=True, blank=True)
    sum_toilet_common = models.BigIntegerField(null=True, blank=True)
    avg_toilet_common = models.FloatField(null=True, blank=True)
    sum_toilet_boys = models.BigIntegerField(null=True, blank=True)
    avg_toilet_boys = models.FloatField(null=True, blank=True)
    sum_toilet_girls = models.BigIntegerField(null=True, blank=True)
    avg_toilet_girls = models.FloatField(null=True, blank=True)
    sum_kitchen_devices_grant = models.BigIntegerField(null=True, blank=True)
    avg_kitchen_devices_grant = models.FloatField(null=True, blank=True)
    sum_has_mdm = models.BigIntegerField(null=True, blank=True)
    sum_has_cal_lab = models.BigIntegerField(null=True, blank=True)
    sum_has_separate_room_for_headmaster = models.BigIntegerField(null=True, blank=True)
    sum_has_electricity = models.BigIntegerField(null=True, blank=True)
    sum_has_boundary_wall = models.BigIntegerField(null=True, blank=True)
    sum_has_library = models.BigIntegerField(null=True, blank=True)
    sum_books_in_library = models.BigIntegerField(null=True, blank=True)
    avg_books_in_library = models.FloatField(null=True, blank=True)
    sum_has_playground = models.BigIntegerField(null=True, blank=True)
    sum_has_blackboard = models.BigIntegerField(null=True, blank=True)
    sum_has_drinking_water = models.BigIntegerField(null=True, blank=True)
    sum_has_medical_checkup = models.BigIntegerField(null=True, blank=True)
    sum_has_ramps = models.BigIntegerField(null=True, blank=True)
    sum_has_computer = models.BigIntegerField(null=True, blank=True)
    sum_has_toilet = models.BigIntegerField(null=True, blank=True)
    sum_no_of_computers = models.BigIntegerField(null=True, blank=True)
    avg_no_of_computers = models.FloatField(null=True, blank=True)
    sum_male_tch = models.BigIntegerField(null=True, blank=True)
    avg_male_tch = models.FloatField(null=True, blank=True)
    sum_female_tch = models.BigIntegerField(null=True, blank=True)
    avg_female_tch = models.FloatField(null=True, blank=True)
    sum_noresp_tch = models.BigIntegerField(null=True, blank=True)
    avg_noresp_tch = models.FloatField(null=True, blank=True)
    sum_head_teacher = models.BigIntegerField(null=True, blank=True)
    avg_head_teacher = models.FloatField(null=True, blank=True)
    sum_graduate_teachers = models.BigIntegerField(null=True, blank=True)
    avg_graduate_teachers = models.FloatField(null=True, blank=True)
    sum_tch_with_professional_qualification = models.BigIntegerField(null=True, blank=True)
    avg_tch_with_professional_qualification = models.FloatField(null=True, blank=True)
    sum_days_involved_in_non_tch_assgn = models.BigIntegerField(null=True, blank=True)
    avg_days_involved_in_non_tch_assgn = models.FloatField(null=True, blank=True)
    sum_teachers_involved_in_non_tch_assgn = models.BigIntegerField(null=True, blank=True)
    avg_teachers_involved_in_non_tch_assgn = models.FloatField(null=True, blank=True)
    sum_boys = models.BigIntegerField(blank=True, null=True)
    avg_boys = models.FloatField(blank=True, null=True)
    sum_girls = models.BigIntegerField(blank=True, null=True)
    avg_girls = models.FloatField(blank=True, null=True)

    objects = models.GeoManager()

    class Meta:
        abstract = True

    @property
    def session(self):
        """Finds session of the current object from table name
        """
        # extracts the 4 digit session from table name
        session = filter(str.isdigit, str(self._meta.db_table))
        listified_session = list(session)
        listified_session.insert(2, '-')
        session = ''.join(listified_session)
        return session


class ClusterAggregations(AggregationBase):
    cluster_name = models.CharField(max_length=50)
    block_name = models.CharField(max_length=50, blank=True)
    district = models.CharField(max_length=50, blank=True)

    entity_type = 'cluster'

    class Meta:
        abstract = True
        unique_together = ("cluster_name", "block_name")

    @property
    def popup_content(self):
        return self.cluster_name

    def schools(self, session=settings.DEFAULT_SESSION):
        SchoolModel = get_models(session, 'school')
        return SchoolModel.objects.filter(
            cluster_name__iexact=self.cluster_name,
            block_name__iexact=self.block_name,
            district__iexact=self.district
        )


class BlockAggregations(AggregationBase):
    block_name = models.CharField(max_length=50)
    district = models.CharField(max_length=50, blank=True)

    entity_type = 'block'

    class Meta:
        abstract = True

    @property
    def popup_content(self):
        return self.block_name

    def schools(self, session=settings.DEFAULT_SESSION):
        SchoolModel = get_models(session, 'school')
        return SchoolModel.objects.filter(
            block_name__iexact=self.block_name,
            district__iexact=self.district
        )


class DistrictAggregations(AggregationBase):
    district = models.CharField(max_length=35)

    class Meta:
        abstract = True

    entity_type = 'district'

    @property
    def popup_content(self):
        return self.district

    @cached_property
    def district_name(self):
        return self.district

    def schools(self, session=settings.DEFAULT_SESSION):
        SchoolModel = get_models(session, 'school')
        return SchoolModel.objects.filter(
            district__iexact=self.district
        )


class AssemblyAggregations(AggregationBase):
    assembly_name = models.CharField(max_length=35)

    entity_type = 'assembly'

    class Meta:
        abstract = True

    @property
    def popup_content(self):
        return self.assembly_name

    def schools(self, session=settings.DEFAULT_SESSION):
        SchoolModel = get_models(session, 'school')
        return SchoolModel.objects.filter(
            assembly_name__iexact=self.assembly_name
        )


class ParliamentAggregations(AggregationBase):
    parliament_name = models.CharField(max_length=35)

    entity_type = 'parliament'

    class Meta:
        abstract = True

    @property
    def popup_content(self):
        return self.parliament_name

    def schools(self, session=settings.DEFAULT_SESSION):
        SchoolModel = get_models(session, 'school')
        return SchoolModel.objects.filter(
            parliament_name__iexact=self.parliament_name
        )


class PincodeAggregations(AggregationBase):
    pincode = models.IntegerField()

    entity_type = 'pincode'

    class Meta:
        abstract = True

    @property
    def popup_content(self):
        return self.pincode

    def schools(self, session=settings.DEFAULT_SESSION):
        SchoolModel = get_models(session, 'school')
        return SchoolModel.objects.filter(
            pincode__iexact=self.pincode
        )


class Dise1011AssemblyAggregations(AssemblyAggregations):
    class Meta:
        db_table = 'dise_1011_assembly_aggregations'


class Dise1112AssemblyAggregations(AssemblyAggregations):
    class Meta:
        db_table = 'dise_1112_assembly_aggregations'


class Dise1213AssemblyAggregations(AssemblyAggregations):
    class Meta:
        db_table = 'dise_1213_assembly_aggregations'


class Dise1314AssemblyAggregations(AssemblyAggregations):
    class Meta:
        db_table = 'dise_1314_assembly_aggregations'


class Dise1415AssemblyAggregations(AssemblyAggregations):
    class Meta:
        db_table = 'dise_1415_assembly_aggregations'


class Dise1011BlockAggregations(BlockAggregations):
    class Meta:
        db_table = 'dise_1011_block_aggregations'


class Dise1112BlockAggregations(BlockAggregations):
    class Meta:
        db_table = 'dise_1112_block_aggregations'


class Dise1213BlockAggregations(BlockAggregations):
    class Meta:
        db_table = 'dise_1213_block_aggregations'


class Dise1314BlockAggregations(BlockAggregations):
    class Meta:
        db_table = 'dise_1314_block_aggregations'


class Dise1415BlockAggregations(BlockAggregations):
    class Meta:
        db_table = 'dise_1415_block_aggregations'


class Dise1011ClusterAggregations(ClusterAggregations):
    class Meta:
        db_table = 'dise_1011_cluster_aggregations'


class Dise1112ClusterAggregations(ClusterAggregations):
    class Meta:
        db_table = 'dise_1112_cluster_aggregations'


class Dise1213ClusterAggregations(ClusterAggregations):
    class Meta:
        db_table = 'dise_1213_cluster_aggregations'


class Dise1314ClusterAggregations(ClusterAggregations):
    class Meta:
        db_table = 'dise_1314_cluster_aggregations'


class Dise1415ClusterAggregations(ClusterAggregations):
    class Meta:
        db_table = 'dise_1415_cluster_aggregations'


class Dise1011ParliamentAggregations(ParliamentAggregations):
    class Meta:
        db_table = 'dise_1011_parliament_aggregations'


class Dise1112ParliamentAggregations(ParliamentAggregations):
    class Meta:
        db_table = 'dise_1112_parliament_aggregations'


class Dise1213ParliamentAggregations(ParliamentAggregations):
    class Meta:
        db_table = 'dise_1213_parliament_aggregations'


class Dise1314ParliamentAggregations(ParliamentAggregations):
    class Meta:
        db_table = 'dise_1314_parliament_aggregations'


class Dise1415ParliamentAggregations(ParliamentAggregations):
    class Meta:
        db_table = 'dise_1415_parliament_aggregations'


class Dise1011DistrictAggregations(DistrictAggregations):
    class Meta:
        db_table = 'dise_1011_district_aggregations'


class Dise1112DistrictAggregations(DistrictAggregations):
    class Meta:
        db_table = 'dise_1112_district_aggregations'


class Dise1213DistrictAggregations(DistrictAggregations):
    class Meta:
        db_table = 'dise_1213_district_aggregations'


class Dise1314DistrictAggregations(DistrictAggregations):
    class Meta:
        db_table = 'dise_1314_district_aggregations'


class Dise1415DistrictAggregations(DistrictAggregations):
    class Meta:
        db_table = 'dise_1415_district_aggregations'


class Dise1011PincodeAggregations(PincodeAggregations):
    class Meta:
        db_table = 'dise_1011_pincode_aggregations'


class Dise1112PincodeAggregations(PincodeAggregations):
    class Meta:
        db_table = 'dise_1112_pincode_aggregations'


class Dise1213PincodeAggregations(PincodeAggregations):
    class Meta:
        db_table = 'dise_1213_pincode_aggregations'


class Dise1314PincodeAggregations(PincodeAggregations):
    class Meta:
        db_table = 'dise_1314_pincode_aggregations'


class Dise1415PincodeAggregations(PincodeAggregations):
    class Meta:
        db_table = 'dise_1415_pincode_aggregations'


class Dise1011BasicData(BasicData):
    class Meta:
        db_table = 'dise_1011_basic_data'


class Dise1112BasicData(BasicData):
    class Meta:
        db_table = 'dise_1112_basic_data'


class Dise1213BasicData(BasicData):
    class Meta:
        db_table = 'dise_1213_basic_data'


class Dise1314BasicData(BasicData):
    class Meta:
        db_table = 'dise_1314_basic_data'


class Dise1415BasicData(BasicData):
    class Meta:
        db_table = 'dise_1415_basic_data'


def get_models(session='10-11', what='all'):
    session = session.replace('-', '')
    schools = __import__('schools')
    models = collections.OrderedDict()

    models['school'] = getattr(
        schools.models, 'Dise{}BasicData'.format(session)
    )
    models['cluster'] = getattr(
        schools.models, 'Dise{}ClusterAggregations'.format(session)
    )
    models['block'] = getattr(
        schools.models, 'Dise{}BlockAggregations'.format(session)
    )
    models['district'] = getattr(
        schools.models, 'Dise{}DistrictAggregations'.format(session)
    )
    models['pincode'] = getattr(
        schools.models, 'Dise{}PincodeAggregations'.format(session)
    )
    models['assembly'] = getattr(
        schools.models, 'Dise{}AssemblyAggregations'.format(session)
    )
    models['parliament'] = getattr(
        schools.models, 'Dise{}ParliamentAggregations'.format(session)
    )

    if what == 'all':
        return models.values()
    else:
        return models.get(what, None)
